src/ring_theory/witt_vector/is_poly.lean

```diff
@@ -0,0 +1,627 @@
+  apply eval₂_hom_congr rfl _ rfl,
+  ext ⟨i, k⟩, fin_cases i;
+  simp only [matrix.head_cons, aeval_X, matrix.cons_val_zero, matrix.cons_val_one],
+end
+
+namespace tactic
+open tactic
+
+/-!
+### The `@[is_poly]` attribute
+
+This attribute is used to derive specialized composition instances
+for `is_poly` and `is_poly₂` declarations.
+-/
+
+/--
+If `n` is the name of a lemma with opened type `∀ vars, tp`,
+where `tp` is an application of `is_poly`,
+`mk_poly_comp_lemmas n vars tp` adds composition instances to the environment
+`n.comp_i` and `n.comp₂_i`.
+-/
+meta def mk_poly_comp_lemmas (n : name) (vars : list expr) (tp : expr) : tactic unit :=
+do c ← mk_const n,
+   `(is_poly %%p _) ← return tp,
+   let appd := vars.foldl expr.app c,
+
+   tgt_bod ← to_expr ``(λ f [hf : is_poly %%p f], is_poly.comp %%appd hf) >>=
+     replace_univ_metas_with_univ_params,
+   tgt_bod ← lambdas vars tgt_bod,
+   tgt_tp ← infer_type tgt_bod,
+   let nm := n <.> "comp_i",
+   add_decl $ mk_definition nm tgt_tp.collect_univ_params tgt_tp tgt_bod,
+   set_attribute `instance nm,
+
+   tgt_bod ← to_expr ``(λ f [hf : is_poly₂ %%p f], is_poly.comp₂ %%appd hf) >>=
+     replace_univ_metas_with_univ_params,
+   tgt_bod ← lambdas vars tgt_bod,
+   tgt_tp ← infer_type tgt_bod,
+   let nm := n <.> "comp₂_i",
+   add_decl $ mk_definition nm tgt_tp.collect_univ_params tgt_tp tgt_bod,
+   set_attribute `instance nm
+
+/--
+If `n` is the name of a lemma with opened type `∀ vars, tp`,
+where `tp` is an application of `is_poly₂`,
+`mk_poly₂_comp_lemmas n vars tp` adds composition instances to the environment
+`n.comp₂_i` and `n.comp_diag`.
+-/
+meta def mk_poly₂_comp_lemmas (n : name) (vars : list expr) (tp : expr) : tactic unit :=
+do c ← mk_const n,
+   `(is_poly₂ %%p _) ← return tp,
+   let appd := vars.foldl expr.app c,
+
+   tgt_bod ← to_expr ``(λ {f g} [hf : is_poly %%p f] [hg : is_poly %%p g],
+     is_poly₂.comp %%appd hf hg) >>= replace_univ_metas_with_univ_params,
+   tgt_bod ← lambdas vars tgt_bod,
+   tgt_tp ← infer_type tgt_bod >>= simp_lemmas.mk.dsimplify,
+   let nm := n <.> "comp₂_i",
+   add_decl $ mk_definition nm tgt_tp.collect_univ_params tgt_tp tgt_bod,
+   set_attribute `instance nm,
+
+   tgt_bod ← to_expr ``(λ {f g} [hf : is_poly %%p f] [hg : is_poly %%p g],
+     (is_poly₂.comp %%appd hf hg).diag) >>= replace_univ_metas_with_univ_params,
+   tgt_bod ← lambdas vars tgt_bod,
+   tgt_tp ← infer_type tgt_bod >>= simp_lemmas.mk.dsimplify,
+   let nm := n <.> "comp_diag",
+   add_decl $ mk_definition nm tgt_tp.collect_univ_params tgt_tp tgt_bod,
+   set_attribute `instance nm
+
+/--
+The `after_set` function for `@[is_poly]`. Calls `mk_poly(₂)_comp_lemmas`.
+-/
+meta def mk_comp_lemmas (n : name) : tactic unit :=
+do d ← get_decl n,
+   (vars, tp) ← open_pis d.type,
+   if tp.is_app_of ``is_poly then
+     mk_poly_comp_lemmas n vars tp
+   else if tp.is_app_of ``is_poly₂ then
+     mk_poly₂_comp_lemmas n vars tp
```

Comment:
If I understand it correctly, this pattern-match is repeated above. Wouldn't it be clearer in the form
```lean
match tp with
`(is_poly %%p _) := mk_poly_comp_lemmas n vars p
...
```
(as the `mk_poly_comp_lemmas` don't use `tp` other than extracting `p`)?